client_code/kompot/_persist.py

```diff
@@ -0,0 +1,248 @@
+            object.__setattr__(self, name, value)
+
+    @property
+    def synced(self):
+        """Whether there are any pending changes recorded."""
+        return not self._changed_attrs
+
+    def clear(self):
+        """Clear any pending changes."""
+        for attr in self._changed_attrs:
+            delattr(self, attr)
+        self._changed_attrs = set()
+
+
+class LinkedAttribute:
+    """A descriptor class for adding linked table items as attributes
+
+    For a class backed by a data tables row, this class is used to dyamically add
+    linked table items as attributes to the parent object.
+    """
+
+    def __init__(self, name, linked_table, linked_attr):
+        """
+        Parameters
+        ----------
+        name: str
+            The name which should be used for the dynamic attribute which is added
+        linked_table: str
+            The name of the column in the row object which links to another table
+        linked_attr: str
+            The name of the column in the linked table which contains the required
+            value
+        """
+        self._name = name
+        self._linked_table = linked_table
+        self._linked_attr = linked_attr
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            raise AttributeError(f"{objtype.__name__} class has no such attribute")
+
+        if instance._row and instance._despatcher.synced:
+            return instance._row[self._linked_table][self._linked_attr]
+        else:
+            return getattr(instance._despatcher, self._linked_attr, None)
+
+    def __set__(self, instance, value):
+        instance._set_value(self._name, value)
+
+
+class RowBackedStore:
+    """A class to hold a data tables row as the persistence mechanism for an object"""
+
+    def __init__(self, server_functions, linked_attributes, row=None, despatcher=None):
+        """
+        Parameters
+        ----------
+        server_functions: dict
+            mapping the keys 'create', 'get', 'update' and 'delete' to the name of the
+            relevant server function for that action
+        linked_attributes: dict
+            mapping the name of a linked column to a further dict which then maps the
+            required name for the dynamically created attribute to the name of the
+            relevant column in the linked table
+        row: anvil.tables.Row
+        despatcher: any
+            A portable class suitable for sending changes from client to server
+        """
+        self._initialised = False
+        self._row = row
+        self._despatcher = despatcher or Despatcher()
+        self._server_functions = server_functions
+        for linked_table, attributes in linked_attributes.items():
+            for attr, linked_attr in attributes.items():
+                setattr(
+                    self.__class__,
+                    attr,
+                    LinkedAttribute(attr, linked_table, linked_attr),
```

Comment:
Are we going to create an issue here if we have conflicts?

Reply:
Yes - not got around to handling that yet!